[FEATURE] Add option to merge categories in categorized renderer

nyalldawson opened the pull request:
This allows users to select multiple existing categories and group them into a single category, which applies to any of the values from the selection.

This allows simpler styling of layers with a large number of categories, where it may be possible to group numerous distinct categories into a smaller, more managable set of categories which apply to multiple values.

Sponsored by SMEC/SJ

![peek 2019-01-07 16-51](https://user-images.githubusercontent.com/1829991/50753128-78a83b00-129c-11e9-916a-dec2cf847305.gif)

nyalldawson committed:
Dox++

nyalldawson committed:
Fix a clazy warning

nyalldawson committed:
Q_FOREACH -> for

nyalldawson committed:
[API] Allow categorized renderer category values to be lists of values

When a category is a list of values, that category will apply to
any of the values from the list

nyalldawson committed:
Add missing documentation for QgsCategorizedRenderer

nyalldawson committed:
[FEATURE] Add option to merge categories in categorized renderer

This allows users to select multiple existing categories and
group them into a single category, which applies to any of the
values from the selection.

This allows simpler styling of layers with a large number of
categories, where it may be possible to group numerous distinct
categories into a smaller, more managable set of categories
which apply to multiple values.

The option is available from the right click context menu
in the categories list view, whenever multiple categories
are selected.

Sponsored by SMEC/SJ

nyalldawson commented:
Fun QGIS Trivia: for a long time we had a button in the categorized renderer which looked like it did this, but actually did nothing...  until @jef-n removed it in https://github.com/qgis/QGIS/commit/812d5b11832c31a875048ceda6609814b194a5ee#diff- dd0bb421d82f2dcfcd950a91befbf5c1 :man_shrugging:

nirvn commented:
Nice one. Are merged categories properly converted to rules when switching to the rule based renderer?

nyalldawson committed:
Don't allow null entry to be merged

nyalldawson committed:
Add option to unmerge categories

nyalldawson committed:
Fix travis

nyalldawson commented:
@m-kuhn I've now moved this option to the context menu and removed the dedicated button. To me it seems more natural to select a bunch of entries and right click, "merge". Plus, less UI bloat.

nyalldawson committed:
Unit tests for unmerging

nyalldawson committed:
Rename some methods for consistency/clarity

nyalldawson committed:
Correctly handle conversion of merged categories to rule based renderer

nyalldawson committed:
Nicer tooltips for merged categories

nyalldawson committed:
More user friendly display of 'all other values' category

nyalldawson commented:
> Are merged categories properly converted to rules when switching to the rule based renderer?

They are now, in fa37e0e

nirvn commented:
+1 to merge, really nifty feature to have.

nyalldawson commented:
Thanks @nirvn!
